### DISCUSSION TITLE
How to generate d.ts files in the same directory as the compiled js files instead of generating them separately in a separate folder

### DISCUSSION THREAD

**Discussion by NexxLuo** [web-infra-dev/rslib#550] [Q&A] [answered]
大部分组件库的d.ts文件和最终的js文件在同级目录中（如下图antd），rslib中默认生成到了独立的文件夹中，尝试配置`tsconfig.json`中的`outDir:"."`但是会报错。该如何处理
![image](https://github.com/user-attachments/assets/3ece5170-9b5e-42ba-b2a1-a4af13d27f76)

复现仓库 : https://github.com/NexxLuo/rslib-demo.git

现象如下：
1. package-a中的d.ts生成到了package-a/lib/src中
2. package-b中的d.ts生成到了package-b/lib/package-b/src中

![image](https://github.com/user-attachments/assets/d4ee5ff9-ab24-44c5-8e2d-58b5cf0e3179)

**Comment by Timeless0911** [comment]
Please provide a [minimal reproduction](https://stackoverflow.com/help/minimal-reproducible-example) to help us address the issue, we cannot assist with troubleshooting based on the information currently provided.
Thanks.

[Why reproduction is required](https://antfu.me/posts/why-reproductions-are-required)

**Reply by NexxLuo** [reply]
已更新复现仓库地址

**Reply by Timeless0911** [reply, marked as answer]
You should add `compilerOptions.rootDir` and `include` field to `tsconfig.json`. It is equivalent to `npx tsc --outDir lib --declarati
on`.

```diff
{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
+   "rootDir": "src",
    "isolatedModules": true,
    "declaration": true
  },
+ "include": ["src"]
}

```

**Reply by NexxLuo** [reply]
配置了rootDir后ts会报以下错误。
![image](https://github.com/user-attachments/assets/28f92aac-81b6-447c-ad88-aec8405499c2)

**Reply by Timeless0911** [reply]
Monorepo 中 workspace 的引用不应该通过 tsconfig.json 的 paths 转换为源码引入，跨包的源码引入已经更改了 tsc 的作用范围，如果要保持每个包打包结果的独立性，应该引用的是打包后的产物，通过在 `packageB` 中声明依赖 `packageA`，通过 `import { ComponentA } from "package-a"`进行引入。
